### PR TITLE
simulate bulk withdrawal

### PR DESCRIPTION
**Pull request by samparsky**
```
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 173010, numberOfEarners: 11, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 159524, numberOfEarners: 17, proofSize: 5 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158014, numberOfEarners: 16, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158024, numberOfEarners: 11, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 159522, numberOfEarners: 18, proofSize: 5 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158004, numberOfEarners: 15, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158004, numberOfEarners: 12, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158014, numberOfEarners: 16, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158004, numberOfEarners: 14, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158014, numberOfEarners: 16, proofSize: 4 
---------------------------------------------------------------------------------
```
Total gas used across 10 channels: 1,440,130

### CODE REVIEW

File: contracts/mocks/Token.sol [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: agpl-3.0
```
Comment (Ivshti) [inline contracts/mocks/Token.sol:1]: you can use the token mock already available in tests/

File: test/simulate.js [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+			)
+			const txns = [
+				await zeroFeeTx(
+					id.address,
+					idInterface.functions.channelOpen.encode([coreAddr, channel.toSolidityTuple()]),
+					0,
+					id,
+					token
+				),
+				await zeroFeeTx(
```
Comment (Ivshti) [inline test/simulate.js:143]: why do we need this?

File: test/simulate.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+			})
+
+			const withdrawSigs = splitSig(await ethSign(channelWithdrawTx.hashHex(), userAcc))
+			const withdrawRoutineReceipt = await (await id.execute(
+				[channelWithdrawTx.toSolidityTuple()],
+				[withdrawSigs],
+				{ gasLimit }
+			)).wait()
+
+			logIdentityExecuteGasInfo(earnerAddresses.length, withdrawRoutineReceipt.gasUsed, proof)
```
Comment (Ivshti) [inline test/simulate.js:193]: this seems like it's only testing individual channels, where's the bulk test?

**Review by Ivshti — CHANGES_REQUESTED** [review]

File: contracts/mocks/Token.sol (hunk 1 above)
Comment (samparsky) [reply contracts/mocks/Token.sol:1]: Truffle has a bug that it doesn't compile `.sol` files in `tests/` dir when I specify a file in the `truffle test /file` command. That's I moved the `mocks` folder into `contracts` dir

**Comment by Ivshti** [comment]
results from a full simulation before v5:

```

  Contract: Simulate Bulk Withdrawal

------- Single Withdrawal w/o identity - channelWithdraw() --------
Gas used: 132588
-------------------------------------------------------

    ✓ routines: open a channel, execute w/o identity: channelWithdraw (1531ms)
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 159526, numberOfEarners: 17, proofSize: 5 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158004, numberOfEarners: 14, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158014, numberOfEarners: 16, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 159536, numberOfEarners: 19, proofSize: 5 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158048, numberOfEarners: 15, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 156536, numberOfEarners: 14, proofSize: 3 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 156548, numberOfEarners: 20, proofSize: 3 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158050, numberOfEarners: 11, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 156526, numberOfEarners: 12, proofSize: 3 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 158028, numberOfEarners: 14, proofSize: 4 
---------------------------------------------------------------------------------

------- Single Channel Bulk Withdrawal  - Identity.execute() --------
Total gas used: 1578816
---------------------------------------------------------------------

    ✓ routines: open a channel, execute: channelWithdraw (7217ms)

------- Bulk Withdrawal - Identity.execute() --------
Total gas used: 1897584
-------------------------------------------------------

    ✓ routines: open a channel, execute bulk: channelWithdraw (5917ms)

```

**Comment by Ivshti** [comment]
And tihs is with 0 fees:

```
  Contract: Simulate Bulk Withdrawal

------- Single Withdrawal w/o identity - channelWithdraw() --------
Gas used: 129786
-------------------------------------------------------

    ✓ routines: open a channel, execute w/o identity: channelWithdraw (1075ms)
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 146383, numberOfEarners: 18, proofSize: 5 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 144839, numberOfEarners: 11, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 146395, numberOfEarners: 17, proofSize: 5 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 146393, numberOfEarners: 20, proofSize: 5 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 146345, numberOfEarners: 18, proofSize: 5 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 144849, numberOfEarners: 11, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 144881, numberOfEarners: 11, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 144897, numberOfEarners: 14, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 146373, numberOfEarners: 19, proofSize: 5 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 144873, numberOfEarners: 13, proofSize: 4 
---------------------------------------------------------------------------------

------- Single Channel Bulk Withdrawal  - Identity.execute() --------
Total gas used: 1456228
---------------------------------------------------------------------

    ✓ routines: open a channel, execute: channelWithdraw (6585ms)

------- Bulk Withdrawal - Identity.execute() --------
Total gas used: 1885905
-------------------------------------------------------

    ✓ routines: open a channel, execute bulk: channelWithdraw (5239ms)
```

only ~13k difference between Identity and non-identity calls, as predicted

**Comment by Ivshti** [comment]
Just single tests with the new OUTPACE (v5):

at commit e6593c0e1cd776e8001d916cb479d132e486ce93 it seems we have consistent ~20k savings

```
------- Single Withdrawal w/o identity - channelWithdraw() --------
Gas used: 111503
-------------------------------------------------------
```

**Comment by Ivshti** [comment]
Changes done:

* solidity optimizer enabled (~7k savings)
* signaturevalidator optimized (~1k savings)

```
------- Single Withdrawal w/o identity - channelWithdraw() --------
Gas used: 102933
```

And the next test with zero fees (slightly changed the label, rounds and no fees):
```
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 112072, numberOfEarners: 15, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 112046, numberOfEarners: 16, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 112048, numberOfEarners: 15, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 112080, numberOfEarners: 14, proofSize: 4 
---------------------------------------------------------------------------------
---------------------------- Gas Information -----------------------------------
Identity.execute() GasUsed: 112036, numberOfEarners: 11, proofSize: 4 
---------------------------------------------------------------------------------

------- Single Channel Identity Withdrawal - new channel every round - Identity.execute() --------
Total gas used: 560282
```

Or an adjusted average of 146k for v5 vs 112k for v5

**Comment by Ivshti** [comment]
Final results for v5, although those are not comparable to v4 as the simulate in there opened a new channel on each round:

```
------- Withdrawal Multiple via identity (10 transactions) - Identity.execute() --------
Total gas used: 589967
-------------------------------------------------------

    ✓ open a channel, execute many withdrawals with identity - same channel (2591ms)

------- Withdrawal Multiple via bulk --------
Total gas used: 374529
-------------------------------------------------------
```

in both cases, zero fees and using the same channel

**Pull request merged**